htlcswitch/link.go

```diff
@@ -1812,6 +1812,33 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 	switch msg := msg.(type) {
 
 	case *lnwire.UpdateAddHTLC:
+		if l.IsFlushing(Incoming) {
+			// This is forbidden by the protocol specification.
+			// The best chance we have to deal with this is to drop
+			// the connection. This should roll back the channel
+			// state to the last CommitSig. If the remote has
+			// already sent a CommitSig we haven't received yet,
+			// channel state will be re-synchronized with a
+			// ChannelReestablish message upon reconnection and the
+			// protocol state that caused us to flush the link will
+			// be rolled back. In the event that there was some
+			// non-deterministic behavior in the remote that caused
+			// them to violate the protocol, we have a decent shot
+			// at correcting it this way, since reconnecting will
+			// put us in the cleanest possible state to try again.
+			l.fail(
```

Comment:
This may be hit erroneously for lnd nodes due to a race:

- counterparty sends `update_add_htlc`, `commit_sig`, `shutdown` in that order
- `update_add_htlc` & `commit_sig` added to link's mailbox
- `shutdown` is processed in `handleCloseMsg` and calls `link.DisableAdds(htlcswitch.Incoming)`
- If `link.DisableAdds` is run before the link can pull from the mailbox, the htlc will cause this section to be hit

The issue here boils down to enforcing the same message ordering that was received over the wire.

Reply:
Where we landed on this is that since we just disconnect, the cost of this race condition being hit is not disastrous. Unless we are prepared to synchronize all message orderings across state machines in this PR, which, while a good idea, is likely beyond the scope of what we want to do here, I think we can leave this as is

Reply:
I'll add an issue and ref it in a comment.